openfisca_core/entities.py

```diff
@@ -14,10 +20,109 @@ class Entity(object):
+        self.entities_json = entities_json
+        self.count = len(entities_json)
+        self.step_size = self.count  # Related to axes.
+        self.ids = sorted(entities_json.keys())
+        for entity_id, entity_object in entities_json.iteritems():
+            check_type(entity_object, dict, [self.plural, entity_id])
+            if not self.is_person:
+                roles_json, variables_json = self.split_variables_and_roles_json(entity_object)
+                self.build_roles(roles_json, entity_id)
+            else:
+                variables_json = entity_object
+            self.build_variables(variables_json, entity_id)
+
+        # Due to set_input mechanism, we must bufferize all inputs, then actually set them, so that the months are set first and the years last.
+        self.finalize_variables_init()
+
+    def build_variables(self, entity_object, entity_id):
+        entity_index = self.ids.index(entity_id)
+        for variable_name, variable_values in entity_object.iteritems():
+            try:
+                self.check_variable_defined_for_entity(variable_name)
+            except ValueError as e:  # The variable is defined for another entity
+                raise SituationParsingError([self.plural, entity_id, variable_name], e.message)
+            except VariableNotFound as e:  # The variable doesn't exist
+                raise SituationParsingError([self.plural, entity_id, variable_name], e.message, code = 404)
+
+            holder = self.get_holder(variable_name)
+
+            if not isinstance(variable_values, dict):
+                raise SituationParsingError([self.plural, entity_id, variable_name],
+                    'Invalid type: must be of type object. Input variables must be set for specific periods. For instance: {"salary": {"2017-01": 2000, "2017-02": 2500}}')
+
+            for date, value in variable_values.iteritems():
+                try:
+                    period = make_period(date)
+                except ValueError as e:
+                    raise SituationParsingError([self.plural, entity_id, variable_name, date], e.message)
+                if value is not None:
+                    array = holder.buffer.get(period)
+                    if array is None:
+                        array = holder.default_array()
+
+                    try:
+                        array[entity_index] = value
+                    except (ValueError, TypeError) as e:
+                        raise SituationParsingError([self.plural, entity_id, variable_name, date],
+                    'Invalid type: must be of type {}.'.format(holder.column.json_type))
+
+                    holder.buffer[period] = array
+
+    def finalize_variables_init(self):
+        for variable_name, holder in self._holders.iteritems():
+            periods = holder.buffer.keys()
+            # We need to handle small periods first for set_input to work
+            sorted_periods = sorted(periods, cmp = compare_period_size)
```

Comment:
Good catch, I was not handling `ETERNITY` !
Should be good now (testes in `test_variables`)